pkg/kubelet/secret/secret_manager.go

```diff
@@ -0,0 +1,246 @@
+limitations under the License.
+*/
+
+package secret
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	apierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/v1"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/clientset"
+	storageetcd "k8s.io/kubernetes/pkg/storage/etcd"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Manager interface {
+	// Get secret by secret namespace and name.
+	GetSecret(namespace, name string) (*v1.Secret, error)
+
+	// WARNING: Register/UnregisterPod functions are not thread safe!
+	// RegisterPod registers all secrets from a given pod.
+	RegisterPod(pod *v1.Pod)
+
+	// UnregisterPod unregisters secrets from a given pod that are not
+	// used by any other registered pod.
+	UnregisterPod(pod *v1.Pod)
+}
+
+// simpleSecretManager implements SecretManager interfaces with
+// simple operations to apiserver.
+type simpleSecretManager struct {
+	kubeClient clientset.Interface
+}
+
+func NewSimpleSecretManager(kubeClient clientset.Interface) (Manager, error) {
+	return &simpleSecretManager{kubeClient: kubeClient}, nil
+}
+
+func (s *simpleSecretManager) GetSecret(namespace, name string) (*v1.Secret, error) {
+	return s.kubeClient.Core().Secrets(namespace).Get(name, metav1.GetOptions{})
+}
+
+func (s *simpleSecretManager) RegisterPod(pod *v1.Pod) {
+}
+
+func (s *simpleSecretManager) UnregisterPod(pod *v1.Pod) {
+}
+
+type objectKey struct {
+	namespace string
+	name      string
+}
+
+// secretStoreItems is a single item stored in secretStore.
+type secretStoreItem struct {
+	sync.Mutex
+
+	secret         *v1.Secret
+	lastUpdateTime time.Time
+	err            error
+	refCount       int
+}
+
+// secretStore is a local cache of secrets.
+type secretStore struct {
+	kubeClient clientset.Interface
+
+	lock             sync.Mutex
+	items            map[objectKey]*secretStoreItem
+	refreshFrequency time.Duration
+}
+
+func newSecretStore(kubeClient clientset.Interface, refreshFrequency time.Duration) *secretStore {
+	return &secretStore{
+		kubeClient:       kubeClient,
+		items:            make(map[objectKey]*secretStoreItem),
+		refreshFrequency: refreshFrequency,
+	}
+}
+
+func isSecretNewer(left, right *v1.Secret) bool {
+	if left != nil && right != nil {
+		newVersion, _ := storageetcd.Versioner.ObjectResourceVersion(left)
+		oldVersion, _ := storageetcd.Versioner.ObjectResourceVersion(right)
+		if newVersion <= oldVersion {
+			return false
+		}
+	}
+	return true
+}
+
+func (s *secretStore) Add(namespace, name string) {
+	key := objectKey{namespace: namespace, name: name}
+	unixEpoch := time.Unix(0, 0)
+	item := func() *secretStoreItem {
+		s.lock.Lock()
+		defer s.lock.Unlock()
+		tmpItem := s.items[key]
+		if tmpItem == nil {
+			s.items[key] = &secretStoreItem{secret: nil, err: nil, lastUpdateTime: unixEpoch, refCount: 1}
+		}
+		return tmpItem
+	}()
+	if item != nil {
+		item.Lock()
+		defer item.Unlock()
```

Comment:
This is not acceptable....SyncLoop calls `RegisterPod`, which calls this `Add` method, but the lock could be held while performing a network operation in `Get`. 

A couple of solutions off the top of my head:
 1. The simple TTL cache with invalidation on ADD that I suggested earlier.
 2. @wojtek-t's background refreshing thread (sorry for the back-and-forth).
 3. use `s.lock` to guard non-network-dependent operations (ref counts, invalidation), and use the per-entry lock for the network operation. 

 ```go
type secretStoreItem struct {
	refCount       int
    secret         *secretTuple
 }
type secretTuple struct {
 	sync.Mutex
 
 	secret         *v1.Secret
 	lastUpdateTime time.Time
 	err            error
}
```
`Add()` will increment the refcount and set `secretStoreItem.secret` to nil. `Get()` will operate mainly on the `secretStoreItem.secret` object.